### ISSUE TITLE
networkd: UseMTU causes networkd to crash and not set MTU

### ISSUE THREAD

**Issue by ajeddeloh** [systemd/systemd#6593]
### Submission type

 - Bug report

### systemd version the issue has been seen with

> 234

### Used distribution

> Container Linux

### In case of bug report: Expected behaviour you didn't see

> The network interface is configured with the MTU from the DHCP lease

### In case of bug report: Unexpected behaviour you saw

> The network interface is configured with the new MTU. It goes down to configure itself. networkd then handles the link going down and sets the MTU back to the original MTU. The link comes back up and networkd crashes.

### In case of bug report: Steps to reproduce the problem

> 

1.  Have a network card that goes down and comes back up when the mtu changes (e1000 will do this)
2.  Set UseMTU to true in the [DHCP] section of a network file.
3.  Have a DHCP server which changes the mtu to 9000
4.  Start networkd
5.  Networkd will crash when the interface comes up from reconfiguring

I think what's happening is when networkd gets the lease with the MTU option it sends a rtnl message to configure the iface to the new mtu asynchonously, then sends the message to configure the iface's address, also asynchonously, with the callback `dhcp4_address_handler`. The iface then goes down to reconfigure. networkd notices this, and tears down all of the state (including the lease structure) associated with that link. The interface then comes up, and the kernel replies to the message to configure the address, triggering the `dhcp4_address_handler` callback. This then calls `link_set_dhcp_routes` which asserts the lease structure is non-NULL and crashes.

I have a [fix](https://github.com/systemd/systemd/pull/6594) that prevents crashing by returning if the lease or network is NULL instead of asserting it, but networkd will still fail to ever set the MTU if the interface goes down to reconfigure.

To actually fix the problem, we could set a timeout when the mtu changes. The link will then go down, come back up, and get a new lease which will not set the MTU (since it will be the same). When the timer expires, if the interface hasn't gotten a new lease, reset the MTU to the old one. The interface will go down, but come back up with the correct MTU, but then the second time it gets a lease, it will already have the correct MTU and not go down again. Does that seem reasonable?

See also: 
 * https://github.com/coreos/bugs/issues/1827 (has detailed repro steps)
 * https://github.com/coreos/bugs/issues/2031

**Comment by poettering** [comment]
@ajeddeloh Do you possibly have a backtrace?

@ssahani maybe something for you?

**Comment by ajeddeloh** [comment]
Here is the backtrace provided by https://github.com/coreos/bugs/issues/2031
```
Program terminated with signal SIGABRT, Aborted.
#0  0x00007f858b508f47 in __GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:54
54	../sysdeps/unix/sysv/linux/raise.c: No such file or directory.
(cros-gdb) bt
#0  0x00007f858b508f47 in __GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:54
#1  0x00007f858b50a4ca in __GI_abort () at abort.c:89
#2  0x00007f858bd09991 in log_assert_failed (text=text@entry=0x5652b24a02a0 "link->dhcp_lease", 
    file=file@entry=0x5652b249fde8 "../systemd-9999/src/network/networkd-dhcp4.c", line=line@entry=61, 
    func=func@entry=0x5652b24a04f0 <__PRETTY_FUNCTION__.14866> "link_set_dhcp_routes") at ../systemd-9999/src/basic/log.c:796
#3  0x00005652b2469f1f in link_set_dhcp_routes (link=0x5652b44312b0) at ../systemd-9999/src/network/networkd-dhcp4.c:61
#4  dhcp4_address_handler (rtnl=0x5652b44295f0, m=0x5652b4433d70, userdata=0x5652b44312b0) at ../systemd-9999/src/network/networkd-dhcp4.c:285
#5  0x00007f858bda61f7 in process_reply (m=<optimized out>, rtnl=0x5652b44295f0) at ../systemd-9999/src/libsystemd/sd-netlink/sd-netlink.c:351
#6  process_running (ret=0x0, rtnl=0x5652b44295f0) at ../systemd-9999/src/libsystemd/sd-netlink/sd-netlink.c:406
#7  sd_netlink_process (rtnl=0x5652b44295f0, ret=ret@entry=0x0) at ../systemd-9999/src/libsystemd/sd-netlink/sd-netlink.c:436
#8  0x00007f858bda65f2 in io_callback (s=<optimized out>, fd=<optimized out>, revents=<optimized out>, userdata=<optimized out>)
    at ../systemd-9999/src/libsystemd/sd-netlink/sd-netlink.c:730
#9  0x00007f858bda1df6 in source_dispatch (s=s@entry=0x5652b4429700) at ../systemd-9999/src/libsystemd/sd-event/sd-event.c:2272
#10 0x00007f858bda357d in sd_event_dispatch (e=e@entry=0x5652b4427c10) at ../systemd-9999/src/libsystemd/sd-event/sd-event.c:2631
#11 0x00007f858bda3752 in sd_event_run (e=e@entry=0x5652b4427c10, timeout=timeout@entry=18446744073709551615) at ../systemd-9999/src/libsystemd/sd-event/sd-event.c:2690
#12 0x00007f858bda38ff in sd_event_loop (e=0x5652b4427c10) at ../systemd-9999/src/libsystemd/sd-event/sd-event.c:2710
#13 0x00005652b24593e0 in main (argc=<optimized out>, argv=<optimized out>) at ../systemd-9999/src/network/networkd.c:143
```

**Comment by poettering** [comment]
Ah, I missed you already had prepped a patch for this

**Comment by ajeddeloh** [comment]
Note the patch only prevents networkd from crashing in this case, it doesn't solve the underlying problem that networkd reverts the link to its old MTU when it sees it went down and lost its dhcp lease: https://github.com/systemd/systemd/blob/master/src/network/networkd-dhcp4.c#L246
The hard part is in most situations that is the right thing to do is in fact reset the MTU when the lease is lost, it just also needs some way to tell it went down due to the MTU changing and not for other reasons.

**Comment by ssahani** [comment]
I reproduced this . I think it's a bug of very worst kind. Also it's exposed a network security flaw. I got in to a *LOOP*  while reproducing this . Once the MTU changed then the DHCP client talking with DHCP server never stops. networkd gets into a loop and generates endless DHCP requests.  Imagine a bunch of machine doing the same thing and can bring down the network. 

@poettering we need to disable `UseMTU=` for this.  I think we should write this in the man also. This triggers a vulnerability .
